File: storage/src/main/java/io/delta/storage/DelegatingLogStore.java
/*
 * Copyright (2021) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.delta.storage;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import io.delta.storage.internal.LogStoreUtils;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileStatus;
import org.apache.hadoop.fs.Path;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class DelegatingLogStore extends LogStore {

    //////////////////////////////////////
    // Static Helper Fields and Methods //
    //////////////////////////////////////

    private static final Logger LOG = LoggerFactory.getLogger(DelegatingLogStore.class);

    public static final String DEFAULT_S3_LOG_STORE_CLASS_NAME =
        S3SingleDriverLogStore.class.getName();

    public static final String DEFAULT_AZURE_LOG_STORE_CLASS_NAME =
        AzureLogStore.class.getName();

    public static final String DEFAULT_HDFS_LOG_STORE_CLASS_NAME = HDFSLogStore.class.getName();

    public static final Set<String> S3_SCHEMES =
            new HashSet<>(Arrays.asList("s3", "s3a", "s3n"));

    public static final Set<String> AZURE_SCHEMES =
            new HashSet<>(Arrays.asList("abfs", "abfss", "adl", "wasb", "wasbs"));

    public static final Set<String> ALL_SCHEMES =
        Stream.concat(S3_SCHEMES.stream(), AZURE_SCHEMES.stream()).collect(Collectors.toSet());

    public static String getDefaultLogStoreClassName(String scheme) {
        if (S3_SCHEMES.contains(scheme)) {
            return DEFAULT_S3_LOG_STORE_CLASS_NAME;
        } else if (AZURE_SCHEMES.contains(scheme)) {
            return DEFAULT_AZURE_LOG_STORE_CLASS_NAME;
        }
        return DEFAULT_HDFS_LOG_STORE_CLASS_NAME;
    }

    //////////////////////////////////
    // Constructor & Helper Methods //
    //////////////////////////////////

    // Map scheme to the corresponding LogStore resolved and created. Accesses to this map needs
    // synchronization. This could be accessed by multiple threads because it is shared through
    // shared DeltaLog instances.
    private final Map<String, LogStore> schemaToLogStoreMap;

    public DelegatingLogStore(Configuration initHadoopConf) {
        super(initHadoopConf);

        this.schemaToLogStoreMap = new HashMap<>();
    }

    public LogStore getDelegateByScheme(Path path, Configuration hadoopConf) throws IOException {
        final String origScheme = path.toUri().getScheme();

        synchronized (this) {
            final String scheme = origScheme == null ? null : origScheme.toLowerCase(Locale.ROOT);
            if (schemaToLogStoreMap.containsKey(scheme)) {
                return schemaToLogStoreMap.get(scheme);
            } else {
                final String className;

                if (scheme == null) {
                    className = DEFAULT_HDFS_LOG_STORE_CLASS_NAME;
                } else {
                    // Resolve LogStore class based on the following order:
                    // 1. Scheme conf if set.
                    // 2. Defaults for scheme if exists.
                    // 3. HDFS log store (implicitly handled by `getDefaultLogStoreClassName`).
                    className = hadoopConf.get(
                            LogStoreUtils.logStoreSchemeConfKey(scheme),
                            getDefaultLogStoreClassName(scheme));
                }

                final LogStore logStore;
                try {
                     logStore = LogStoreUtils.createLogStoreWithClassName(className, hadoopConf);
                } catch (ClassNotFoundException | NoSuchMethodException |
                        InvocationTargetException | InstantiationException |
                        IllegalAccessException e) {
                    // This method is used by all DelegatingLogStore public APIs, which each throw
                    // IOException. It's not worth changing the parent LogStore interface and
                    // exposing this specific implementation Exception types, so we catch and wrap
                    // those exceptions, instead.
                    throw new IOException("DelegatingLogStore: internal error", e);
                }

                schemaToLogStoreMap.put(scheme, logStore);

                LOG.info(
                    String.format(
                        "LogStore %s is used for scheme %s", logStore.getClass().getName(), scheme
                    )
                );

                return logStore;
            }
        }
    }

    //////////////////////////
    // Public API Overrides //
    //////////////////////////

    @Override
    public CloseableIterator<String> read(Path path, Configuration hadoopConf) throws IOException {
        return getDelegateByScheme(path, hadoopConf).read(path, hadoopConf);
    }

    @Override
    public void write(
            Path path,
            Iterator<String> actions,
            Boolean overwrite,
            Configuration hadoopConf) throws IOException {
        getDelegateByScheme(path, hadoopConf).write(path, actions, overwrite, hadoopConf);
    }

    @Override
    public Iterator<FileStatus> listFrom(Path path, Configuration hadoopConf) throws IOException {
        return getDelegateByScheme(path, hadoopConf).listFrom(path, hadoopConf);
    }

    @Override
    public Path resolvePathOnPhysicalStorage(
            Path path,
            Configuration hadoopConf) throws IOException {
        return getDelegateByScheme(path, hadoopConf).resolvePathOnPhysicalStorage(path, hadoopConf);
    }

    @Override
    public Boolean isPartialWriteVisible(Path path, Configuration hadoopConf) throws IOException {
        return getDelegateByScheme(path, hadoopConf).isPartialWriteVisible(path, hadoopConf);
    }
}

File: standalone/src/main/scala/io/delta/standalone/internal/storage/DelegatingLogStore.scala
/*
 * Copyright (2020-present) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.delta.standalone.internal.storage

import java.util.Locale

import scala.collection.mutable

import io.delta.storage.{CloseableIterator, LogStore}
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.{FileStatus, Path}

import io.delta.standalone.internal.logging.Logging

/**
 * A delegating LogStore used to dynamically resolve LogStore implementation based
 * on the scheme of paths.
 */
class DelegatingLogStore(hadoopConf: Configuration)
  extends LogStore(hadoopConf)
  with Logging {

  // Map scheme to the corresponding LogStore resolved and created. Accesses to this map need
  // synchronization This could be accessed by multiple threads because it is shared through
  // shared DeltaLog instances.
  private val schemeToLogStoreMap = mutable.Map.empty[String, LogStore]

  private lazy val defaultLogStore = createLogStore(DelegatingLogStore.defaultHDFSLogStoreClassName)

  // Creates a LogStore with given LogStore class name.
  private def createLogStore(className: String): LogStore = {
    LogStoreProvider.createLogStoreWithClassName(className, hadoopConf)
  }

  // Create LogStore based on the scheme of `path`.
  private def schemeBasedLogStore(path: Path): LogStore = {
    Option(path.toUri.getScheme) match {
      case Some(origScheme) =>
        val scheme = origScheme.toLowerCase(Locale.ROOT)
        this.synchronized {
          if (schemeToLogStoreMap.contains(scheme)) {
            schemeToLogStoreMap(scheme)
          } else {
            // Resolve LogStore class based on the following order:
            // 1. Scheme conf if set.
            // 2. Defaults for scheme if exists.
            // 3. Default.
            val logStoreClassNameOpt = Option(
              hadoopConf.get(LogStoreProvider.logStoreSchemeConfKey(scheme))
            ).orElse(DelegatingLogStore.getDefaultLogStoreClassName(scheme))

            val logStore = logStoreClassNameOpt.map(createLogStore(_)).getOrElse(defaultLogStore)
            schemeToLogStoreMap += scheme -> logStore
            logInfo(s"LogStore ${logStore.getClass.getName} is used for scheme ${scheme}")
            logStore
          }
        }
      case _ => defaultLogStore
    }
  }

  def getDelegate(path: Path): LogStore = schemeBasedLogStore(path)

  //////////////////////////
  // Public API Overrides //
  //////////////////////////

  override def read(path: Path, hadoopConf: Configuration): CloseableIterator[String] = {
    getDelegate(path).read(path, hadoopConf)
  }

  override def write(
      path: Path,
      actions: java.util.Iterator[String],
      overwrite: java.lang.Boolean,
      hadoopConf: Configuration): Unit = {
    getDelegate(path).write(path, actions, overwrite, hadoopConf)
  }

  override def listFrom(path: Path, hadoopConf: Configuration): java.util.Iterator[FileStatus] = {
    getDelegate(path).listFrom(path, hadoopConf)
  }

  override def resolvePathOnPhysicalStorage(path: Path, hadoopConf: Configuration): Path = {
    getDelegate(path).resolvePathOnPhysicalStorage(path, hadoopConf)
  }

  override def isPartialWriteVisible(path: Path, hadoopConf: Configuration): java.lang.Boolean = {
    getDelegate(path).isPartialWriteVisible(path, hadoopConf)
  }
}

object DelegatingLogStore {

  /**
   * Java LogStore (io.delta.storage) implementations are now the default.
   */
  val defaultS3LogStoreClassName = classOf[io.delta.storage.S3SingleDriverLogStore].getName
  val defaultAzureLogStoreClassName = classOf[io.delta.storage.AzureLogStore].getName
  val defaultHDFSLogStoreClassName = classOf[io.delta.storage.HDFSLogStore].getName
  val defaultGCSLogStoreClassName = classOf[io.delta.storage.GCSLogStore].getName

  // Supported schemes with default.
  val s3Schemes = Set("s3", "s3a", "s3n")
  val azureSchemes = Set("abfs", "abfss", "adl", "wasb", "wasbs")
  val gsSchemes = Set("gs")

  // Returns the default LogStore class name for `scheme`.
  // None if we do not have a default for it.
  def getDefaultLogStoreClassName(scheme: String): Option[String] = {
    if (s3Schemes.contains(scheme)) {
      return Some(defaultS3LogStoreClassName)
    } else if (DelegatingLogStore.azureSchemes(scheme: String)) {
      return Some(defaultAzureLogStoreClassName)
    } else if (DelegatingLogStore.gsSchemes(scheme: String)) {
      return Some(defaultGCSLogStoreClassName)
    }
    None
  }
}

